README.md
[![NPM version](https://badge.fury.io/js/cdk-efs-assets.svg)](https://badge.fury.io/js/cdk-efs-assets)
[![PyPI version](https://badge.fury.io/py/cdk-efs-assets.svg)](https://badge.fury.io/py/cdk-efs-assets)
![Release](https://github.com/pahud/cdk-efs-assets/workflows/Release/badge.svg)

# cdk-efs-assets

CDK construct library to populate Amazon EFS assets from Github or S3. If the source is S3, the construct also optionally supports updating the contents in EFS if a new zip file is uploaded to S3.

## Install

TypeScript/JavaScript:

```bash
npm i cdk-efs-assets
```

## SyncedAccessPoint

The main construct that is used to provide this EFS sync functionality is `SyncedAccessPoint`. This extends the standard EFS `AccessPoint` construct, and takes an additional `SyncSource` constructor property which defines the source to sync assets from. The `SyncedAccessPoint` instance can be used anywhere an `AccessPoint` can be used. For example, to specify a volume in a Task Definition:

```ts
const taskDefinition = new ecs.FargateTaskDefinition(this, 'TaskDefinition', {
  ... 
  volumes: [
    {
      name: 'efs-storage',
      efsVolumeConfiguration: {
        fileSystemId: sharedFileSystem.fileSystemId,
        transitEncryption: 'ENABLED',
        authorizationConfig: {
          accessPointId: syncedAccessPoint.accessPointId
        }
      }
    },
  ]
});
```

## Sync Engine

This library supports both `AWS Fargate` and `AWS Lambda` as the sync engine. As AWS Lambda currently has know issue with Amazon EFS([#100](https://github.com/pahud/cdk-efs-assets/issues/100)), the default sync engine is `AWS Fargate`. You can opt in AWS Lambda with the `engine` construct property of `SyncedAccessPoint`.


## Sync Source


Use `GithubSyncSource` and `S3ArchiveSyncSource` construct classes to define your `syncSource` from Github
or Amazon S3 bucket. For example:

To define a public github repository as the `syncSource`:

```ts
new SyncedAccessPoint(stack, 'EfsAccessPoint', {
  ...
  syncSource: new GithubSyncSource({
    vpc,
    repository: 'https://github.com/pahud/cdk-efs-assets.git',
  }),
});
```

To define a private github repository as the `syncSource`:

```ts
new SyncedAccessPoint(stack, 'EfsAccessPoint', {
  ...
  syncSource: new GithubSyncSource({
    vpc,
    repository: 'https://github.com/pahud/private-repo.git',
    secret: {
      id: 'github',
      key: 'oauth_token',
    },
  }),
});
```


### `syncDirectoryPath`

By default, the synced EFS assets are placed into a directory corresponding to the type of the sync source. For example, the default behavior of the GitHub source is to place the copied files into a directory named the same as the repository name (for a repository specified as 'https://github.com/pahud/cdk-efs-assets.git', the directory name would be 'cdk-efs-assets'), while the default behavior of the S3 archive source is to place the copied files into a directory named the same as the zip file (for a zip file name of 'assets.zip', the directory name would be 'assets').

If you wish to override this default behavior, specify a value for the `syncDirectoryPath` property that is passed into the `SyncSource` call.

If you are using the `AccessPoint` in an ECS/Fargate Task Definition, you probably will want to override the value of `syncDirectoryPath` to '/'. This will place the file contents in the root directory of the Access Point. The reason for this is that when you create a volume that is referencing an EFS Access Point, you are not allowed to specify any path other than the root directory in the task definition configuration.

## How to use SyncedAccessPoint initialized with files provisioned from GitHub repository

This will sync assets from a GitHub repository to a directory (by default, the output directory is named after the repository name) in the EFS AccessPoint:

```ts
import { SyncSource, SyncedAccessPoint } from 'cdk-efs-assets';

const app = new App();

const env = {
  region: process.env.CDK_DEFAULT_REGION ?? AWS_DEFAULT_REGION,
  account: process.env.CDK_DEFAULT_ACCOUNT,
};

const stack = new Stack(app, 'testing-stack', { env });

const vpc = ec2.Vpc.fromLookup(stack, 'Vpc', { isDefault: true })

const fileSystem = new efs.FileSystem(stack, 'Filesystem', {
  vpc,
  removalPolicy: RemovalPolicy.DESTROY,
})

const efsAccessPoint = new SyncedAccessPoint(stack, 'GithubAccessPoint', {
  vpc,
  fileSystem,
  path: '/demo-github',
  createAcl: {
    ownerGid: '1001',
    ownerUid: '1001',
    permissions: '0755',
  },
  posixUser: {
    uid: '1001',
    gid: '1001',
  },
  syncSource: new GithubSyncSource({
    vpc,
    repository: 'https://github.com/pahud/cdk-efs-assets.git',
  })
});
```

### Github private repository support

To clone a github private repository, you need to generate your github **PAT(Personal Access Token)** and store the token in **AWS Secrets Manager** secret.

For example, if your PAT is stored in the AWS Secret manager with the secret ID `github` and the key `oauth_token`, you can specify the `secret` property as the sample below. Under the covers the lambda function will format the full github repository uri with your **PAT** and successfully git clone the private repository to the efs filesystem.

Store your PAT into the AWS Secrets Manager with AWS CLI:

```sh
aws secretsmanager create-secret \
--name github \
--secret-string '{"oauth_token":"MYOAUTHTOKEN"}'
```

Configure the `secret` property to allow lambda to retrieve the **PAT** from the secret:


```ts
new GithubSyncSource({
    vpc,
    repository: 'https://github.com/username/repo.git',
    secret: {
      id: 'github',
      key: 'oauth_token',
    },
})
```

## How to use SyncedAccessPoint initialized with files provisioned from zip file stored in S3

This will sync assets from a zip file stored in an S3 bucket to a directory (by default, the output directory is named after the zip file name) in the EFS AccessPoint:

```ts
import { S3ArchiveSync } from 'cdk-efs-assets';

const app = new App();

const env = {
  region: process.env.CDK_DEFAULT_REGION ?? AWS_DEFAULT_REGION,
  account: process.env.CDK_DEFAULT_ACCOUNT,
};

const stack = new Stack(app, 'testing-stack', { env });

const vpc = ec2.Vpc.fromLookup(stack, 'Vpc', { isDefault: true })

const fileSystem = new efs.FileSystem(stack, 'Filesystem', {
  vpc,
  removalPolicy: RemovalPolicy.DESTROY,
})

const bucket = Bucket.fromBucketName(this, 'Bucket', 'demo-bucket');

const efsAccessPoint = new SyncedAccessPoint(stack, 'EfsAccessPoint', {
  vpc,
  fileSystem,
  path: '/demo-s3',
  createAcl: {
    ownerGid: '1001',
    ownerUid: '1001',
    permissions: '0755',
  },
  posixUser: {
    uid: '1001',
    gid: '1001',
  },
  syncSource: new S3ArchiveSyncSource({
    vpc,
    bucket,
    zipFilePath: 'folder/foo.zip',
  }),
});
```

### syncOnUpdate

If the `syncOnUpdate` property is set to `true` (defaults to `true`), then the specified zip file path will be monitored, and if a new object is uploaded to the path, then it will resync the data to EFS. Note that to use this functionality, you must have a CloudTrail Trail in your account that captures the desired S3 write data event.

This feature is only available with the `LAMBDA` sync engine.

*WARNING*: The contents of the extraction directory in the access point will be destroyed before extracting the zip file.


# `StatefulFargateNginx`

This library comes with `StatefulFargateNginx` construct which allows you to build an Amazon EFS-backed stateful
AWS Fargate service with its document root seeded from any github repository.

See this [tweet](https://twitter.com/pahudnet/status/1367792169063354371) for the demo.

Sample:

```ts
new StatefulFargateNginx(this, 'NyanCat', {
  vpc,
  github: 'https://github.com/cristurm/nyan-cat.git',
});

```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.3.47](https://github.com/pahud/cdk-efs-assets/compare/v0.3.46...v0.3.47) (2021-04-21)

### [0.3.46](https://github.com/pahud/cdk-efs-assets/compare/v0.3.45...v0.3.46) (2021-04-20)

### [0.3.45](https://github.com/pahud/cdk-efs-assets/compare/v0.3.44...v0.3.45) (2021-04-19)

### [0.3.44](https://github.com/pahud/cdk-efs-assets/compare/v0.3.43...v0.3.44) (2021-04-18)

### [0.3.43](https://github.com/pahud/cdk-efs-assets/compare/v0.3.42...v0.3.43) (2021-04-17)

### [0.3.42](https://github.com/pahud/cdk-efs-assets/compare/v0.3.41...v0.3.42) (2021-04-16)

### [0.3.41](https://github.com/pahud/cdk-efs-assets/compare/v0.3.40...v0.3.41) (2021-04-12)

### [0.3.40](https://github.com/pahud/cdk-efs-assets/compare/v0.3.39...v0.3.40) (2021-04-11)

### [0.3.39](https://github.com/pahud/cdk-efs-assets/compare/v0.3.38...v0.3.39) (2021-04-10)

### [0.3.38](https://github.com/pahud/cdk-efs-assets/compare/v0.3.37...v0.3.38) (2021-04-09)

### [0.3.37](https://github.com/pahud/cdk-efs-assets/compare/v0.3.36...v0.3.37) (2021-04-08)

### [0.3.36](https://github.com/pahud/cdk-efs-assets/compare/v0.3.35...v0.3.36) (2021-04-07)

### [0.3.35](https://github.com/pahud/cdk-efs-assets/compare/v0.3.34...v0.3.35) (2021-04-05)

### [0.3.34](https://github.com/pahud/cdk-efs-assets/compare/v0.3.33...v0.3.34) (2021-04-04)

### [0.3.33](https://github.com/pahud/cdk-efs-assets/compare/v0.3.32...v0.3.33) (2021-04-03)

### [0.3.32](https://github.com/pahud/cdk-efs-assets/compare/v0.3.31...v0.3.32) (2021-04-01)

### [0.3.31](https://github.com/pahud/cdk-efs-assets/compare/v0.3.30...v0.3.31) (2021-04-01)

### [0.3.30](https://github.com/pahud/cdk-efs-assets/compare/v0.3.29...v0.3.30) (2021-03-31)

### [0.3.29](https://github.com/pahud/cdk-efs-assets/compare/v0.3.28...v0.3.29) (2021-03-30)

### [0.3.28](https://github.com/pahud/cdk-efs-assets/compare/v0.3.27...v0.3.28) (2021-03-29)

### [0.3.27](https://github.com/pahud/cdk-efs-assets/compare/v0.3.26...v0.3.27) (2021-03-27)

### [0.3.26](https://github.com/pahud/cdk-efs-assets/compare/v0.3.25...v0.3.26) (2021-03-26)

### [0.3.25](https://github.com/pahud/cdk-efs-assets/compare/v0.3.24...v0.3.25) (2021-03-25)

### [0.3.24](https://github.com/pahud/cdk-efs-assets/compare/v0.3.23...v0.3.24) (2021-03-24)

### [0.3.23](https://github.com/pahud/cdk-efs-assets/compare/v0.3.22...v0.3.23) (2021-03-22)

### [0.3.22](https://github.com/pahud/cdk-efs-assets/compare/v0.3.21...v0.3.22) (2021-03-21)

### [0.3.21](https://github.com/pahud/cdk-efs-assets/compare/v0.3.20...v0.3.21) (2021-03-20)

### [0.3.20](https://github.com/pahud/cdk-efs-assets/compare/v0.3.19...v0.3.20) (2021-03-19)

### [0.3.19](https://github.com/pahud/cdk-efs-assets/compare/v0.3.18...v0.3.19) (2021-03-18)

### [0.3.18](https://github.com/pahud/cdk-efs-assets/compare/v0.3.17...v0.3.18) (2021-03-18)

### [0.3.17](https://github.com/pahud/cdk-efs-assets/compare/v0.3.16...v0.3.17) (2021-03-17)

### [0.3.16](https://github.com/pahud/cdk-efs-assets/compare/v0.3.15...v0.3.16) (2021-03-16)

### [0.3.15](https://github.com/pahud/cdk-efs-assets/compare/v0.3.14...v0.3.15) (2021-03-15)

### [0.3.14](https://github.com/pahud/cdk-efs-assets/compare/v0.3.13...v0.3.14) (2021-03-14)

### [0.3.13](https://github.com/pahud/cdk-efs-assets/compare/v0.3.12...v0.3.13) (2021-03-13)

### [0.3.12](https://github.com/pahud/cdk-efs-assets/compare/v0.3.11...v0.3.12) (2021-03-12)

### [0.3.11](https://github.com/pahud/cdk-efs-assets/compare/v0.3.10...v0.3.11) (2021-03-11)

### [0.3.10](https://github.com/pahud/cdk-efs-assets/compare/v0.3.9...v0.3.10) (2021-03-09)

### [0.3.9](https://github.com/pahud/cdk-efs-assets/compare/v0.3.8...v0.3.9) (2021-03-08)

### [0.3.8](https://github.com/pahud/cdk-efs-assets/compare/v0.3.7...v0.3.8) (2021-03-07)

### [0.3.7](https://github.com/pahud/cdk-efs-assets/compare/v0.3.6...v0.3.7) (2021-03-06)

### [0.3.6](https://github.com/pahud/cdk-efs-assets/compare/v0.3.5...v0.3.6) (2021-03-06)

### [0.3.5](https://github.com/pahud/cdk-efs-assets/compare/v0.3.4...v0.3.5) (2021-03-06)

### [0.3.4](https://github.com/pahud/cdk-efs-assets/compare/v0.3.3...v0.3.4) (2021-03-05)


### Features

* add StatefulFargateNginx  ([#115](https://github.com/pahud/cdk-efs-assets/issues/115)) ([459548d](https://github.com/pahud/cdk-efs-assets/commit/459548d06945579f0d052993b0f452f027d64609))

### [0.3.3](https://github.com/pahud/cdk-efs-assets/compare/v0.3.2...v0.3.3) (2021-03-05)

### [0.3.2](https://github.com/pahud/cdk-efs-assets/compare/v0.3.1...v0.3.2) (2021-03-04)

### [0.3.1](https://github.com/pahud/cdk-efs-assets/compare/v0.3.0...v0.3.1) (2021-03-04)

### [0.2.37](https://github.com/pahud/cdk-efs-assets/compare/v0.2.36...v0.2.37) (2021-03-04)

## [0.3.0](https://github.com/pahud/cdk-efs-assets/compare/v0.2.37...v0.3.0) (2021-03-04)

### [0.2.37](https://github.com/pahud/cdk-efs-assets/compare/v0.2.36...v0.2.37) (2021-03-04)


### Features

* refactor with AWS Fargate support ([#113](https://github.com/pahud/cdk-efs-assets/issues/113)) ([894b863](https://github.com/pahud/cdk-efs-assets/commit/894b86397f2793ff6e7b72ed90f3245052ad42f1))

### [0.2.36](https://github.com/pahud/cdk-efs-assets/compare/v0.2.35...v0.2.36) (2021-03-04)

### [0.2.35](https://github.com/pahud/cdk-efs-assets/compare/v0.2.34...v0.2.35) (2021-03-03)

### [0.2.34](https://github.com/pahud/cdk-efs-assets/compare/v0.2.33...v0.2.34) (2021-03-02)

### [0.2.33](https://github.com/pahud/cdk-efs-assets/compare/v0.2.32...v0.2.33) (2021-03-01)

### [0.2.32](https://github.com/pahud/cdk-efs-assets/compare/v0.2.31...v0.2.32) (2021-02-28)

### [0.2.31](https://github.com/pahud/cdk-efs-assets/compare/v0.2.30...v0.2.31) (2021-02-27)

### [0.2.30](https://github.com/pahud/cdk-efs-assets/compare/v0.2.29...v0.2.30) (2021-02-26)


src/integ.default.ts
import * as ec2 from '@aws-cdk/aws-ec2';
import * as efs from '@aws-cdk/aws-efs';
import { Bucket } from '@aws-cdk/aws-s3';
import { App, Stack, RemovalPolicy, Construct } from '@aws-cdk/core';
import { StatefulFargateNginx } from './stateful-fargate';
import { SyncedAccessPoint, GithubSyncSource, S3ArchiveSyncSource } from './synced-access-point';

export class IntegTesting {
  readonly stack: Stack[];

  constructor() {
    const app = new App();

    const env = {
      region: process.env.CDK_DEFAULT_REGION,
      account: process.env.CDK_DEFAULT_ACCOUNT,
    };

    const stack = new Stack(app, 'testing-stack', { env });

    const vpc = getOrCreateVpc(stack);

    const fs = new efs.FileSystem(stack, 'Filesystem', {
      vpc,
      removalPolicy: RemovalPolicy.DESTROY,
    });

    const bucketName = stack.node.tryGetContext('BUCKET_NAME') || 'mock';
    const bucket = Bucket.fromBucketName(stack, 'ImportedBucket', bucketName);

    // checkout the public github repo to efs filesystem
    new SyncedAccessPoint(stack, 'GithubSyncedAccessPoint', {
      vpc,
      fileSystem: fs,
      path: '/demo-github',
      createAcl: {
        ownerGid: '1001',
        ownerUid: '1001',
        permissions: '0755',
      },
      posixUser: {
        uid: '1001',
        gid: '1001',
      },
      syncSource: new GithubSyncSource({
        vpc,
        repository: 'https://github.com/pahud/cdk-efs-assets.git',
      }),
    });

    // checkout the private github repo to efs filesystem
    new SyncedAccessPoint(stack, 'GithubSyncedAccessPointPrivate', {
      vpc,
      fileSystem: fs,
      path: '/demo-github-private',
      createAcl: {
        ownerGid: '1001',
        ownerUid: '1001',
        permissions: '0755',
      },
      posixUser: {
        uid: '1001',
        gid: '1001',
      },
      syncSource: new GithubSyncSource({
        vpc,
        repository: 'https://github.com/pahud/private-repo.git',
        secret: {
          id: 'github',
          key: 'oauth_token',
        },
      }),
    });

    new SyncedAccessPoint(stack, 'S3SyncedAccessPoint', {
      vpc,
      fileSystem: fs,
      path: '/demo-s3-archive',
      createAcl: {
        ownerGid: '1001',
        ownerUid: '1001',
        permissions: '0755',
      },
      posixUser: {
        uid: '1001',
        gid: '1001',
      },
      syncSource: new S3ArchiveSyncSource({
        vpc,
        bucket,
        zipFilePath: 'folder/foo.zip',
      }),
    });

    this.stack = [stack];
  }
}


export class NyanCatDemo {
  readonly stack: Stack[];
  constructor() {
    const app = new App();

    const env = {
      region: process.env.CDK_DEFAULT_REGION,
      account: process.env.CDK_DEFAULT_ACCOUNT,
    };

    const stack = new Stack(app, 'testing-stack', { env });

    const vpc = getOrCreateVpc(stack);

    new StatefulFargateNginx(stack, 'NyanCat', {
      vpc,
      github: 'https://github.com/cristurm/nyan-cat.git',
    });

    this.stack = [stack];
  }
}

// run the integ testing
new IntegTesting();

// run Nyan Cat demo
// new NyanCatDemo();


function getOrCreateVpc(scope: Construct): ec2.IVpc {
  // use an existing vpc or create a new one
  return scope.node.tryGetContext('use_default_vpc') === '1' ?
    ec2.Vpc.fromLookup(scope, 'Vpc', { isDefault: true }) :
    scope.node.tryGetContext('use_vpc_id') ?
      ec2.Vpc.fromLookup(scope, 'Vpc', { vpcId: scope.node.tryGetContext('use_vpc_id') }) :
      new ec2.Vpc(scope, 'Vpc', { maxAzs: 3, natGateways: 1 });
}
